api/inc/register_gateway.h

```diff
@@ -17,104 +17,208 @@
+ *  instruction.
+ * @internal
+ * @note The branch instruction is encoded according to the Thumb-2 immediate
+ * encoding rules:
+ *     <instr_addr>: B.N <label>
+ *     imm = (<label>_addr - PC) / 2
+ * Where:
+ *     PC = <instr>_addr + 4
+ * The +4 is to account for the pipelined PC at the time of the branch
+ * instruction. See ARM DDI 0403E.b page A4-102 for more details.
+ * @param instr[in] Address of the branch instruction
+ * @param label[in] Address of the label
+ * @returns the 16-bit encoding of the B.N <label> instruction.
+ */
+#define BRANCH_OPCODE(instr, label) \
+    (uint16_t) (0xE000 | (uint8_t) ((((uint32_t) (label) - (uint32_t) (instr + 4)) / 2) & 0xFF))
 
-/* 1 argument: simple read, no mask */
-#define __UVISOR_REGISTER_GATEWAY_METADATA1(src_box, addr) \
-    "b.n skip_args%=\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_REGISTER_GATEWAY_MAGIC) "\n" \
-    ".word " UVISOR_TO_STRING(addr) "\n" \
-    ".word " UVISOR_TO_STRING(src_box) "_cfg_ptr\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_OP_READ(UVISOR_OP_NOP)) "\n" \
-    ".word " UVISOR_TO_STRING(__UVISOR_OP_DEFAULT_MASK) "\n" \
-    "skip_args%=:\n"
+/** `BX LR` encoding
+ * @internal
+ */
+#define BXLR                           ((uint16_t) 0x4770)
 
-/* 2 arguments: simple write, no mask */
-#define __UVISOR_REGISTER_GATEWAY_METADATA2(src_box, addr, val) \
-    "b.n skip_args%=\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_REGISTER_GATEWAY_MAGIC) "\n" \
-    ".word " UVISOR_TO_STRING(addr) "\n" \
-    ".word " UVISOR_TO_STRING(src_box) "_cfg_ptr\n" \
-    ".word " UVISOR_TO_STRING(UVISOR_OP_WRITE(UVISOR_OP_NOP)) "\n" \
-    ".word " UVISOR_TO_STRING(__UVISOR_OP_DEFAULT_MASK) "\n" \
-    "skip_args%=:\n"
+/** Register Gateway - Read operation
+ *
+ * This macro provides an API to perform 32-bit read operations on restricted
+ * registers. Such accesses are assembled into a read-only flash structure that
+ * is read and validated by uVisor before performing the operation.
+ *
+ * @warning These APIs currently only support link-time known values. This means
+ * that an access of this type is not supported:
+ * ```
+ * static uint32_t const const_mask = 0x0000F000;
+ * a = uvisor_read(&SYSCFG->CTRL, UVISOR_RGW_OP_READ, const_mask); // Yes
+ * uint32_t var_mask = rand();
+ * a = uvisor_read(&SYSCFG->CTRL, UVISOR_RGW_OP_READ, var_mask); // No
+ * ```
+ *
+ * @param box_name[in]  The name of the source box as decalred in
+ *                      `UVISOR_BOX_CONFIG`.
+ * @param addr[in]      The address for the data access.
+ * @param operation[in] The operation to perform at the address for the read. It
+ *                      is chosen among the `UVISOR_RGW_OP_*` macros.
+ * @param mask[in]      The mask to apply for the read operation.
+ * @returns The value read from address using the operation and mask provided
+ * (or their respective defaults if they have not been provided).
+ */
+#define uvisor_read(box_name, addr, op, msk) \
+    ({ \
```

Comment:
So, we aren't using `offsetof` because of ARMCC. But we are using `({`, even though ARMCC, because ARMCC is not supported. Both these uses are in api/inc headers.